### PR TITLE
[BREAK] Move read-receipts to EE and threads check marks improvements

### PR DESCRIPTION
**Pull request by matheusbsilva137**
## Proposed changes (including videos or screenshots)

- Improvement: add read receipts for messages in threads only when the thread is open;
- Move read-receipts feature to EE (under the `message-read-receipt` module). The `chat.getMessageReadReceipts` endpoint and `getReadReceipts` Meteor method are now EE only.

## Further comments

[TC-73]

[TC-73]: https://rocketchat.atlassian.net/browse/TC-73?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/meteor/server/services/reads/service.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+		const threadMessage = Messages.findOneById(tmid, { projection: { ts: 1, tlm: 1, rid: 1 } });
+		if (!threadMessage || !threadMessage.tlm) {
+			return;
+		}
+
+		ReadReceipt.storeThreadMessagesReadReceipts(tmid, userId, read?.ls || threadMessage.ts);
+
+		// doesn't mark as read if not all room members have read the thread
+		const subscriptions = await Subscriptions.findByRoomId(threadMessage.rid, {
```
Comment (sampaiodiego) [inline apps/meteor/server/services/reads/service.ts:25]: this query should filter out archived subscriptions.. please remember to create a test case for this scenario

File: apps/meteor/server/services/reads/service.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import { Messages, Subscriptions } from '../../../app/models/server';
+import { ReadReceipt } from '../../../imports/message-read-receipt/server/lib/ReadReceipt';
+
+export class ReadsService extends ServiceClassInternal implements IReadsService {
+	protected name = 'reads';
+
+	async readThread(userId: string, tmid: string): Promise<void> {
+		const read = await Reads.findOneByUserIdAndThreadId(userId, tmid);
+
+		await Reads.updateReadTimestampByUserIdAndThreadId(userId, tmid);
```
[the remark this answers is not in the document]
Comment (matheusbsilva137) [reply apps/meteor/server/services/reads/service.ts]: Fixed!

File: apps/meteor/server/services/reads/service.ts (hunk 1 above)
Comment (matheusbsilva137) [reply apps/meteor/server/services/reads/service.ts]: Fixed! I'll now add some tests for this

**Comment by outofrange** [comment]
Since I just stumbled over this, and haven't found any other info about `Move read-receipts feature to EE` - does this mean read notifications on chat messages (a very useful feature) will be removed from the upcoming Community Edition?

**Review by LucianoPierdona — DISMISSED** [review]

**Review by hugocostadev — DISMISSED** [review]

**Comment by milton-rucks** [comment]
> Since I just stumbled over this, and haven't found any other info about `Move read-receipts feature to EE` - does this mean read notifications, those little check marks to the right of each message, a very useful feature, will be removed from the upcoming Community Edition?

Hi @outofrange 

Yes, it’s correct.
We are improving the read receipts on threads and moving the feature to EE.
It has been announced since 2020 as part of the Enterprise offer through our old pricing page, but due to a mistake, it continued to be available on CE.
The official communication will come together with the 6.0 launch.

File: apps/meteor/ee/server/local-services/reads/service.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+		const threadMessage = Messages.findOneById(tmid, { projection: { ts: 1, tlm: 1, rid: 1 } });
+		if (!threadMessage || !threadMessage.tlm) {
+			return;
+		}
+
+		await Reads.updateReadTimestampByUserIdAndThreadId(userId, tmid);
+		ReadReceipt.storeThreadMessagesReadReceipts(tmid, userId, read?.ls || threadMessage.ts);
+
+		// doesn't mark as read if not all room members have read the thread
+		const subscriptions = await Subscriptions.findByRoomId(threadMessage.rid, {
```
Comment (rodrigok) [inline apps/meteor/ee/server/local-services/reads/service.ts:25]: You should retrieve the subscriptions only if they are less then `MAX_ROOM_SIZE_CHECK_INDIVIDUAL_READ_RECEIPTS`. So you can do a count first and retrieve them later, you can exclude the archived ones in the query as well

**Review by rodrigok — CHANGES_REQUESTED** [review]

**Review by LucianoPierdona — DISMISSED** [review]

File: apps/meteor/server/models/raw/Reads.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+	}
+
+	getMinimumLastSeenByThreadId(tmid: IMessage['_id']): Promise<Reads | null> {
+		return this.findOne(
+			{
+				tmid,
+			},
+			{
+				sort: {
+					ls: 1,
```
Comment (rodrigok) [inline apps/meteor/server/models/raw/Reads.ts:27]: A sort without an index is always a bad idea ;)

File: apps/meteor/tests/end-to-end/api/05-chat.js [hunk 5]
```diff
@@ -1346,39 +1346,6 @@ describe('[Chat]', function () {
 		});
 	});
 
-	describe('[/chat.getMessageReadReceipts]', () => {
```
Comment (rodrigok) [inline apps/meteor/tests/end-to-end/api/05-chat.js:1349]: Why were those testes removed?

**Review by rodrigok — CHANGES_REQUESTED** [review]

File: apps/meteor/tests/end-to-end/api/05-chat.js (hunk 5 above)
Comment (LucianoPierdona) [reply apps/meteor/tests/end-to-end/api/05-chat.js:1349]: It's referenced here to add a new file to cover both EE and CE, I believe:
https://github.com/RocketChat/Rocket.Chat/pull/27074/commits/a31375aeeeb33afe1c05aa22cd1f4c757848722d

Since this PR is supposed to go to the 6.0, I don't think it's viable to do these tests today

File: apps/meteor/server/models/raw/Reads.ts (hunk 4 above)
Comment (matheusbsilva137) [reply apps/meteor/server/models/raw/Reads.ts:27]: Index added. I had one index there already, but I guess it's best to keep this as a separate one, right?

File: apps/meteor/tests/end-to-end/api/05-chat.js (hunk 5 above)
Comment (matheusbsilva137) [reply apps/meteor/tests/end-to-end/api/05-chat.js:1349]: Yeah, not only I think this should be in a separate file, but also this needs to be updated since this endpoint is now EE-only. I plan to get the tests I worked on #26611 (old PR related to this issue), but these also need to be changed since this feature has been moved to EE
I'm planning to add this as a regression later so as not to hold this PR

File: apps/meteor/ee/server/local-services/reads/service.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
+import { Reads, Subscriptions } from '@rocket.chat/models';
+import type { ISubscription } from '@rocket.chat/core-typings';
+import { ServiceClassInternal } from '@rocket.chat/core-services';
+
+import type { IReadsService } from '../../sdk/types/IReadsService';
+import { Messages } from '../../../../app/models/server';
+import { ReadReceipt } from '../../lib/message-read-receipt/ReadReceipt';
+import { MAX_ROOM_SIZE_CHECK_INDIVIDUAL_READ_RECEIPTS } from '../../lib/constants';
+
+export class ReadsService extends ServiceClassInternal implements IReadsService {
```
Comment (KevLehman) [inline apps/meteor/ee/server/local-services/reads/service.ts:10]: A bit picky, but could we name this `ReadReceipts`? 
I know `Reads` makes sense when you have the context, but otherwise it's not very descriptive

File: apps/meteor/ee/app/message-read-receipt/server/hooks/afterReadMessages.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
+import type { IUser, IRoom, IMessage } from '@rocket.chat/core-typings';
+
+import { Reads } from '../../../../server/sdk';
```
Comment (KevLehman) [inline apps/meteor/ee/app/message-read-receipt/server/hooks/afterReadMessages.ts:3]: Can't you import it from @rocket.chat/core-services?

File: apps/meteor/ee/app/message-read-receipt/server/index.js [hunk 8]
```diff
@@ -0,0 +1,5 @@
+import { onLicense } from '../../license/server';
+
+onLicense('message-read-receipt', () => {
+	require('./hooks');
```
Comment (KevLehman) [inline apps/meteor/ee/app/message-read-receipt/server/index.js:4]: You can use `await import('....')` to not use require 👀

File: apps/meteor/ee/server/api/chat.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+			try {
+				return API.v1.success({
+					receipts: await Meteor.call('getReadReceipts', { messageId }),
+				});
+			} catch (error: any) {
+				return API.v1.failure({
+					error: error.message,
+				});
+			}
```
Comment (KevLehman) [inline apps/meteor/ee/server/api/chat.ts:30]: Since the error is just being returned, do you really need the try/catch or can we rely on the endpoint's try/catch?

**Review by KevLehman — CHANGES_REQUESTED** [review]

File: apps/meteor/tests/end-to-end/api/05-chat.js (hunk 5 above)
Comment (matheusbsilva137) [reply apps/meteor/tests/end-to-end/api/05-chat.js:1349]: Thanks @LucianoPierdona for adding this :hugs:

File: apps/meteor/ee/server/local-services/reads/service.ts (hunk 6 above)
Comment (matheusbsilva137) [reply apps/meteor/ee/server/local-services/reads/service.ts:10]: I do agree that `Reads` is not a meaningful name, but I also think that calling it `ReadReceipts` could be misleading :thinking: since it manages the `Reads` collection (which currently works like subscriptions to threads, but we plan to expand it)

File: apps/meteor/ee/app/message-read-receipt/server/hooks/afterReadMessages.ts (hunk 7 above)
Comment (matheusbsilva137) [reply apps/meteor/ee/app/message-read-receipt/server/hooks/afterReadMessages.ts]: Done! Thanks for suggesting this, though (I had forgotten to add this service to the `core-services` package :grimacing: )

File: apps/meteor/ee/app/message-read-receipt/server/index.js (hunk 8 above)
Comment (matheusbsilva137) [reply apps/meteor/ee/app/message-read-receipt/server/index.js]: Done! Also converted this file to TS

File: apps/meteor/ee/server/api/chat.ts (hunk 9 above)
Comment (matheusbsilva137) [reply apps/meteor/ee/server/api/chat.ts:30]: Hmmm I see we're doing something similar (the try/catch strategy) in some other endpoints. Not sure about what's the ideal behavior here. Maybe @sampaiodiego can help us here

File: apps/meteor/tests/end-to-end/api/05-chat.js [hunk 10]
```diff
@@ -1347,8 +1347,14 @@ describe('[Chat]', function () {
 	});
 
 	describe('[/chat.getMessageReadReceipts]', () => {
+		const isEnterprise = typeof process.env.IS_EE === 'string' ? process.env.IS_EE === 'true' : !!process.env.IS_EE;
 		describe('when execute successfully', () => {
 			it("should return the statusCode 200 and 'receipts' property and should be equal an array", (done) => {
+				if (!isEnterprise) {
+					done();
```
Comment (rodrigok) [inline apps/meteor/tests/end-to-end/api/05-chat.js:1354]: Would be better to skip it, so we know it's a test that didn't run rather than run successfully

File: apps/meteor/ee/server/local-services/reads/service.ts (hunk 6 above)
Comment (KevLehman) [reply apps/meteor/ee/server/local-services/reads/service.ts:10]: Just out of curiosity, why we needed another collection for thread subscriptions? 👀

File: apps/meteor/tests/end-to-end/api/05-chat.js (hunk 10 above)
Comment (matheusbsilva137) [reply apps/meteor/tests/end-to-end/api/05-chat.js]: Agree, done!

File: apps/meteor/app/threads/server/methods/getThreadMessages.js [hunk 11]
```diff
@@ -28,18 +29,24 @@ Meteor.methods({
 
 		const user = Meteor.user();
 		const room = Rooms.findOneById(thread.rid);
+		callbacks.run('beforeReadMessages', thread.rid, user._id);
```
Comment (sampaiodiego) [inline apps/meteor/app/threads/server/methods/getThreadMessages.js:32]: this should be moved to right before the call to `readThread(..)`, after all checks are done

File: apps/meteor/ee/server/local-services/reads/service.ts (hunk 6 above)
Comment (matheusbsilva137) [reply apps/meteor/ee/server/local-services/reads/service.ts:10]: That's because we decided to treat read receipts for messages in threads just as if threads were "separate rooms". So in order to control which messages have been read, we gotta know whether everyone in the room opened the thread and the messages they read.
For messages in the main room, we use subscriptions' `ls` field (`lastSeen`), so that we can get the minimum value of this field among room members and mark all older messages as read. We needed something similar for threads.

File: apps/meteor/ee/server/api/chat.ts (hunk 9 above)
Comment (sampaiodiego) [reply apps/meteor/ee/server/api/chat.ts:30]: this try/catch is redundant.. as @KevLehman mentioned we already have that for every REST endpoint. as you can see [here](https://github.com/RocketChat/Rocket.Chat/blob/2bea8dc006351de264e1424534f0299098c304fd/apps/meteor/app/api/server/api.js#L419)

File: apps/meteor/server/methods/readThreads.js [hunk 12]
```diff
@@ -22,12 +23,15 @@ Meteor.methods({
 		}
 
 		const user = Meteor.user();
+		callbacks.run('beforeReadMessages', thread.rid, user._id);
```
Comment (sampaiodiego) [inline apps/meteor/server/methods/readThreads.js:26]: same as the other one, this should be moved to right before the call to `readThread()` function

File: packages/core-typings/src/IRoom.ts [hunk 13]
```diff
@@ -74,6 +74,7 @@ export interface IRoom extends IRocketChatRecord {
 	usernames?: string[];
 	ts?: Date;
 
+	closedAt?: Date;
```
Comment (sampaiodiego) [inline packages/core-typings/src/IRoom.ts:77]: why was this moved from `IOmnichannelGenericRoom` to `IRoom`?

**Review by sampaiodiego — CHANGES_REQUESTED** [review]

File: apps/meteor/ee/server/local-services/reads/service.ts (hunk 6 above)
Comment (sampaiodiego) [reply apps/meteor/ee/server/local-services/reads/service.ts:10]: what do you guys think of calling the service as `MessageReads`? a bit more meaningful I guess 😬

File: packages/core-typings/src/IRoom.ts (hunk 13 above)
Comment (matheusbsilva137) [reply packages/core-typings/src/IRoom.ts:77]: That's because I converted [this file](https://github.com/RocketChat/Rocket.Chat/blob/1e5a0650d9afd8c2649765241e14f58a40443f62/apps/meteor/imports/message-read-receipt/server/hooks.js#L14) to TS and I'd have the `"Property closedAt does not exist on type IRoom"` error if I didn't do so. Also, since any room can be closed and have this property, I guess it makes sense to move it to `IRoom`. wdyt?

File: apps/meteor/server/methods/readThreads.js (hunk 12 above)
Comment (matheusbsilva137) [reply apps/meteor/server/methods/readThreads.js]: Thankss, fixed!

File: apps/meteor/ee/server/api/chat.ts (hunk 9 above)
Comment (matheusbsilva137) [reply apps/meteor/ee/server/api/chat.ts]: great, fixed!

File: packages/core-typings/src/IRoom.ts (hunk 13 above)
Comment (sampaiodiego) [reply packages/core-typings/src/IRoom.ts:77]: not really, only omnichannel rooms can be closed.. or am I missing something? how do you close a room that is not an omnichannel room?

File: apps/meteor/app/threads/server/methods/getThreadMessages.js (hunk 11 above)
Comment (matheusbsilva137) [reply apps/meteor/app/threads/server/methods/getThreadMessages.js]: tks, done!

File: apps/meteor/ee/server/local-services/reads/service.ts (hunk 6 above)
Comment (matheusbsilva137) [reply apps/meteor/ee/server/local-services/reads/service.ts:10]: hmmm sound better to me :thinking: it's currently just "threads' reads", but it's better for us not to be so specific, I guess
I'll update the service name

Comment (matheusbsilva137) [reply apps/meteor/ee/server/local-services/reads/service.ts]: Done! Updated the service name to `MessageReads` and the collection name to `message_reads`

File: packages/core-typings/src/IRoom.ts (hunk 13 above)
Comment (matheusbsilva137) [reply packages/core-typings/src/IRoom.ts]: You're right, I guess I confused it with the `closedAt` field in subscriptions :grimacing: 
That's fixed.

**Review by sampaiodiego — APPROVED** [review]

**Review by rodrigok — APPROVED** [review]

**Review by ggazzo — APPROVED** [review]

**Review by KevLehman — APPROVED** [review]

**Review by LucianoPierdona — APPROVED** [review]

**Comment by ae5960e8-a6fc-491f-b252-898ecf59af95** [comment]
Please see this forum post regarding this change:
https://forums.rocket.chat/t/update-6-0-0-read-receipts-complaint/16164/4
